### PR TITLE
[pushbullet] initial action contribution

### PR DESCRIPTION
**Pull request by hakan42**
I have used this action for quite some time locally in my OH2 setup. @kubawolanin reported success, too.

Topic: https://community.openhab.org/t/pushbullet-binding-and-action/3409/34?u=hakan

Signed-off-by: Hakan Tandogan <hakan@tandogan.com>

### CODE REVIEW

**Comment by hakan42** [comment]
Closing and reopening to trigger build

**Comment by hakan42** [comment]
Again, closing and re-opening

**Comment by hakan42** [comment]
Another try with Jenkins...

**Comment by hakan42** [comment]
And yet another try at PR building...

**Comment by hakan42** [comment]
The action worked fine over the last weekend. The build breaker is caused by my attempts to get the PR builder working.

@kaikreuzer , do we have a chance to get this into 2.1 or should it wait for 2.2 ?

**Comment by kaikreuzer** [comment]
1000 loc that aren't reviewed yet - I would suggest to schedule it for the 2.2 :-)

**Comment by kubawolanin** [comment]
I'm totally digging this add-on! :) It'll surely replace my current openhab-pushbullet bot based on node.js.
I'm going to test it this weekend or later next week and report back.
Thanks @hakan42 for your work!

**Comment by hakan42** [comment]
@kubawolanin , did you manage to test the pushbullet action a bit?

**Comment by kubawolanin** [comment]
Hi @hakan42,
Yes, I've been playing around with this and I like it a lot!
Didn't find any flaws so far - works perfectly for me :)

**Comment by hakan42** [comment]
Triggering Jenkins once again...

**Comment by hakan42** [comment]
Dear @1-x-add-ons-maintainers , does anyone have a bit of time for a review? We have at least two successful user of this action ( @kubawolanin and me 😄 )

**Comment by 9037568** [comment]
Glad to start a review of this.

**Comment by hakan42** [comment]
* Fixed since annotations to read 1.11.0
* Fixed years to read -2017

File: bundles/action/org.openhab.action.pushbullet/src/main/java/org/openhab/action/pushbullet/internal/model/PushResponse.java [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+                "active='" + active + '\'' +
+                ", iden='" + iden + '\'' +
+                ", type='" + type + '\'' +
+                ", dismissed=" + dismissed +
+                ", direction='" + direction + '\'' +
+                ", senderIdentifier='" + senderIdentifier + '\'' +
+                ", senderEmail='" + senderEmail + '\'' +
+                ", senderEmailNormalized='" + senderEmailNormalized + '\'' +
+                ", senderName='" + senderName + '\'' +
+                ", receiverIdentifer='" + receiverIdentifer + '\'' +
```
Comment (hakan42) [inline bundles/action/org.openhab.action.pushbullet/src/main/java/org/openhab/action/pushbullet/internal/model/PushResponse.java]: done

File: bundles/action/org.openhab.action.pushbullet/build.properties [hunk 2]
```diff
@@ -0,0 +1,7 @@
+source.. = src/main/java/,\
+           src/main/resources/
+bin.includes = META-INF/,\
+               .,\
+               OSGI-INF/,\
+               about.html
```
Comment (hakan42) [inline bundles/action/org.openhab.action.pushbullet/build.properties]: The OH2 static code checks require about.html to be distributed. Just in preparation for whenever OH1 will be statically checked as well 😄

File: bundles/action/org.openhab.action.pushbullet/src/main/java/org/openhab/action/pushbullet/internal/PushbulletAPIConnector.java [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+     * @return
+     */
+    private static boolean sendPush(String botName, String recipient, String title, String body, String type) {
+        boolean result = false;
+
+        logger.trace("    Botname is   '{}'", botName);
+
+        PushbulletBot bot = bots.get(botName);
+        if (bot == null) {
+            logger.error("Unconfigured pushbullet bot, please check configuration");
```
Comment (hakan42) [inline bundles/action/org.openhab.action.pushbullet/src/main/java/org/openhab/action/pushbullet/internal/PushbulletAPIConnector.java]: agreed, and implemented.

File: bundles/action/org.openhab.action.pushbullet/src/main/java/org/openhab/action/pushbullet/internal/PushbulletAPIConnector.java [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+            String responseString = HttpUtil.executeUrl(METHOD, API_URL_PUSHES, headers, stream, CONTENT_TYPE, TIMEOUT);
+
+            logger.trace("    Got Response: {}", responseString);
+            PushResponse response = gson.fromJson(responseString, PushResponse.class);
+
+            logger.trace("    Unpacked Response: {}", response);
+
+            stream.close();
+
+            if (null == response.getPushError()) {
```
Comment (hakan42) [inline bundles/action/org.openhab.action.pushbullet/src/main/java/org/openhab/action/pushbullet/internal/PushbulletAPIConnector.java]: agreed, and implemented.

File: bundles/action/org.openhab.action.pushbullet/src/main/java/org/openhab/action/pushbullet/internal/PushbulletAPIConnector.java [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+                result = true;
+            }
+        } catch (IOException e) {
+            logger.trace("IO problems pushing note: {}", e.getMessage());
```
Comment (hakan42) [inline bundles/action/org.openhab.action.pushbullet/src/main/java/org/openhab/action/pushbullet/internal/PushbulletAPIConnector.java]: agreed, and implemented.

File: bundles/action/org.openhab.action.pushbullet/src/main/java/org/openhab/action/pushbullet/internal/PushbulletActionService.java [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+import org.osgi.service.cm.ManagedService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static org.openhab.action.pushbullet.internal.PushbulletConstants.ACCESS_TOKEN_KEY;
+import static org.openhab.action.pushbullet.internal.PushbulletConstants.BOTS_KEY;
+import static org.openhab.action.pushbullet.internal.PushbulletConstants.DEFAULT_BOTNAME;
+
+/**
+ * This class...
```
Comment (hakan42) [inline bundles/action/org.openhab.action.pushbullet/src/main/java/org/openhab/action/pushbullet/internal/PushbulletActionService.java]: agreed, and implemented.

File: bundles/action/org.openhab.action.pushbullet/src/main/java/org/openhab/action/pushbullet/internal/PushbulletActionService.java [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+                    PushbulletAPIConnector.addBot(botName, accessToken);
+                }
+            }
+
+            // Now, try to configure the default bot just in case we have older configuration around
+            PushbulletAPIConnector.addBot(DEFAULT_BOTNAME, defaultAccessToken);
+
+            PushbulletAPIConnector.logPushbulletBots();
+
+            // Actually, we should check whether there are more than zero bots configured...
```
Comment (hakan42) [inline bundles/action/org.openhab.action.pushbullet/src/main/java/org/openhab/action/pushbullet/internal/PushbulletActionService.java]: agreed, and implemented.

File: bundles/action/org.openhab.action.pushbullet/src/main/java/org/openhab/action/pushbullet/internal/PushbulletActivator.java [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+ */
+package org.openhab.action.pushbullet.internal;
+
+import org.osgi.framework.BundleActivator;
+import org.osgi.framework.BundleContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class...
```
Comment (hakan42) [inline bundles/action/org.openhab.action.pushbullet/src/main/java/org/openhab/action/pushbullet/internal/PushbulletActivator.java]: agreed, and implemented.

File: bundles/action/org.openhab.action.pushbullet/src/main/java/org/openhab/action/pushbullet/internal/model/PushResponse.java (hunk 1 above)
Comment (hakan42) [inline bundles/action/org.openhab.action.pushbullet/src/main/java/org/openhab/action/pushbullet/internal/model/PushResponse.java]: agreed, and implemented.

File: features/openhab-addons-external/src/main/resources/conf/pushbullet.cfg [hunk 9]
```diff
@@ -0,0 +1,26 @@
+#
+# Pushbullet action configuration
+#
+
+#
+# You can set up multiple accounts if you want to have seperate channels or just go with the
```
Comment (hakan42) [inline features/openhab-addons-external/src/main/resources/conf/pushbullet.cfg]: agreed, and implemented.

**Comment by hakan42** [comment]
> This fails to build. I see Jenkins isn't complaining, but I get many errors like the one below from both Maven and Eclipse.
> 
> [ERROR] Failed to execute goal org.eclipse.tycho:tycho-compiler-plugin:1.0.0:compile (default-compile) on project org.openhab.action.pushbullet: Compilation failure: Compilation failure:
> [ERROR] \openhab\bundles\action\org.openhab.action.pushbullet\src\main\java\org\openhab\action\pushbullet\internal\model\Push.java:[11]
> [ERROR] import com.google.gson.annotations.SerializedName;
> [ERROR] ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> [ERROR] Access restriction: The type 'SerializedName' is not API (restriction on classpath entry '\.m2\repository\p2\osgi\bundle\com.google.gson\2.5.0\com.google.gson-2.5.0.jar')

Actually, I have no idea why this started breaking. When I built the binding, this used to compile both with maven on the command line and on Jenkins + Travis.

Also, my other binding in the OH2 tree, the sensebox binding, compiles flawlessy even though it references the same gson library version.

@kaikreuzer , do you have any idea what might have changed in OH1 since about 1. August 2017 or so?

File: bundles/action/org.openhab.action.pushbullet/build.properties (hunk 2 above)
Comment (kaikreuzer) [inline bundles/action/org.openhab.action.pushbullet/build.properties]: @9037568 It is especially helpful in the binary as this is what is distributed :-)

**Comment by kaikreuzer** [comment]
@hakan42 I think there is a simple answer: You seem to only [import `com.google.gson`](https://github.com/openhab/openhab1-addons/pull/5225/files#diff-62fe44d90fd35254d253c7378d012699R12), but not `com.google.gson.annotations` :-)

**Comment by hakan42** [comment]
> You seem to only import com.google.gson, but not com.google.gson.annotations :-)

Oops 😊 

Fixed.

**Comment by hakan42** [comment]
@9037568 , @kaikreuzer , I fixed all the code issues on August, but could not find the time to work on the README file since then.

Would it be acceptable to you if we merge the plugin itself, so it is available in the next release of OH, and I promise that I will provide a seperate PR for the documentation fixes in the next major holidays (looking at ma calender, this would be end of December) ?

**Comment by 9037568** [comment]
I have no objection unless someone else does.

**Comment by kaikreuzer** [comment]
@hakan42 As there already is at least a README file and all comments on it came from @9037568 and he does not object, I don't have any veto either.
My only suggestion would be that you remove the two "TODO" entries within the README as they would look a bit weird in the published documentation.

File: bundles/action/org.openhab.action.pushbullet/README.md [hunk 10]
```diff
@@ -4,12 +4,19 @@ The Pushbullet action allows you to notify iOS, Android & Windows 10 Phone & Des
...
+For specific information on each item, see the [Pushbullet API](https://docs.pushbullet.com/).
+
+* `sendPushbulletNote(String receiver, String title, String message)`
+* `sendPushbulletNote(String botname, String receiver, String title, String message)`
+
+## Examples
 
 ```
-pushbullet(String botname, String receiver, String title, String message)
+sendPushbulletNote("pitaya", "hakan.tandogan@gmail.com", "this is the title", "And this is the body of the message")
```
Comment (kubawolanin) [inline bundles/action/org.openhab.action.pushbullet/README.md]: Are you sure you want your private email exposed in the docs? :-)

**Comment by hakan42** [comment]
I have removed the TODOs, but Jenkins seems to have a problem again with the squeezebox binding:

```
Missing requirement: org.openhab.action.squeezebox 1.11.0.qualifier requires 'package org.openhab.io.squeezeserver 0.0.0'
```

Could we merge this even with ignoring squeezebox?

File: bundles/action/org.openhab.action.pushbullet/README.md (hunk 10 above)
Comment (hakan42) [reply bundles/action/org.openhab.action.pushbullet/README.md]: aaaah, now you mention it :)

Comment (hakan42) [reply bundles/action/org.openhab.action.pushbullet/README.md]: And removed, thank you for seeing this.

**Comment by hakan42** [comment]
Ah, ignore the squeezebox remark, Jenkins was successful on the second run 😄

**Comment by 9037568** [comment]
Thanks, @hakan42 !

**Pull request merged**